### PR TITLE
Add support for Volunteer Leaderboard, linking Volunteer/Groups/Membership & Integrated with Action Items

### LINKED ISSUE #2585: GSoC: Add Support for Volunteer, Group & Membership
**Is your feature request related to a problem? Please describe.**
- Add support Volunteer, Groups and memberships. 
- Refactor existing queries & mutation to support search/sort functionalities. 
- Add support to integrate with organization action items. 

=

**Potential internship candidates**

Please read this if you are planning to apply for a Palisadoes Foundation internship 
- https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by GlenDsza**
**What kind of change does this PR introduce?**

feature, refactoring

**Issue Number:**

Fixes #2585 

**Did you add tests for your changes?**

Yes 

**Summary**

- Added `VolunteerMembership` Model to handle request/invitation for volunteer groups
- Added relevant resolver (query+mutation) to get membership details for admin and user
- Added getVolunteerRanks query Resolver to add support volunteer leaderboard.
- Updated Action Item to be integrated with Volunteers & Groups
- Added Search/Sort/Filter functionalities.
- Remove unused child resolvers
- Added common check functions for checking existence of an entity and reduce redundant code

**Does this PR introduce a breaking change?**

No 

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes 

## Summary by CodeRabbit

- **New Features**
	- Introduced new GraphQL types and inputs for managing volunteer memberships and ranks.
	- Enhanced action items with new fields for assignee types and groups.
	- Added new queries for retrieving action items by user and event volunteers.
	- Implemented a new mutation for creating volunteer memberships.
	- Expanded functionality for event volunteers and groups, including new fields and improved relationships.
	- Added filtering capabilities for action items based on assignee names and categories.
	- Introduced new queries for fetching volunteer memberships and ranks.
	- Enhanced the `eventsByOrganizationConnection` resolver to include upcoming events filtering.
	- Added support for sorting volunteer ranks and memberships by various criteria.

- **Bug Fixes**
	- Improved error handling for scenarios involving volunteers and groups.

- **Documentation**
	- Updated documentation to reflect changes in data models and resolver functionalities.

- **Tests**
	- Added new test suites for the `getVolunteerMembership` and `getVolunteerRanks` resolvers.
	- Expanded existing tests to cover new functionalities and error handling scenarios.
	- Enhanced tests for action items and event volunteer management.
	- Improved test coverage for the `adminCheck` utility and related functions.
	- Updated tests to reflect changes in data structure and resolver logic.
	- Enhanced tests for the `createActionItem` and `updateVolunteerMembership` mutations.

### CODE REVIEW

**Comment by GlenDsza** [comment]
@tasneemkoushar PTAL

File: src/resolvers/Event/actionItems.ts [hunk 1]
```diff
@@ -15,6 +15,6 @@ import type { EventResolvers } from "../../types/generatedGraphQLTypes";
  */
 export const actionItems: EventResolvers["actionItems"] = async (parent) => {
   return await ActionItem.find({
-    eventId: parent._id,
+    event: parent._id,
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/resolvers/Event/actionItems.ts:18]: event is required as per schema

File: tests/resolvers/Mutation/updateVolunteerMembership.spec.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    const updatedMembership = await updateVolunteerMembership?.(
+      {},
+      { id: membership?._id.toString() ?? "", status: "accepted" },
+      { userId: testUser1?._id },
+    );
+
+    expect(updatedMembership?.status).toEqual("accepted");
+  });
+});
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply tests/resolvers/Mutation/updateVolunteerMembership.spec.ts:75]: Added

File: src/resolvers/Mutation/updateVolunteerMembership.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        },
+      ).lean()) as InterfaceVolunteerMembership;
+
+    // Handle additional updates if the status is accepted
+    if (args.status === "accepted") {
+      await handleAcceptedStatusUpdates(updatedVolunteerMembership);
+    }
+
+    return updatedVolunteerMembership;
+  };
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/resolvers/Mutation/updateVolunteerMembership.ts:95]: done

File: tests/resolvers/Mutation/updateActionItem.spec.ts [hunk 4]
```diff
@@ -171,16 +256,290 @@ describe("resolvers -> Mutation -> updateActionItem", () => {
...
+
+      const context = {
+        userId: testUser?._id,
+      };
+
+      await updateActionItemResolver?.({}, args, context);
+      await updateActionItemResolver?.({}, args2, context);
+    } catch (error: unknown) {
+      console.log(error);
+    }
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply tests/resolvers/Mutation/updateActionItem.spec.ts:346]: removed unnecessary try catch

File: tests/utilities/checks.spec.ts [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+});
+
+afterAll(async () => {
+  await disconnect(MONGOOSE_INSTANCE);
+});
+
+describe("utilities -> checks", () => {
+  afterEach(() => {
+    vi.resetModules();
+  });
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply tests/utilities/checks.spec.ts:66]: done

File: tests/utilities/checks.spec.ts [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+    try {
+      await checkUserExists(testUser?.appUserProfileId);
+    } catch (error: unknown) {
+      expect((error as Error).message).toEqual(
+        `Translated ${USER_NOT_FOUND_ERROR.MESSAGE}`,
+      );
+    }
+    expect(spy).toBeCalledWith(USER_NOT_FOUND_ERROR.MESSAGE);
+  });
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply tests/utilities/checks.spec.ts:81]: done refactored

File: tests/resolvers/Mutation/createActionItem.spec.ts [hunk 7]
```diff
@@ -1,341 +1,243 @@
...
+      event: testEvent?._id,
+      volunteer: testEventVolunteer1._id,
+      status: "accepted",
+    },
+    {
+      event: testEvent?._id,
+      volunteer: testEventVolunteer1._id,
+      status: "rejected",
+    },
+  ]);
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply tests/resolvers/Mutation/createActionItem.spec.ts:78]: done

File: src/resolvers/Mutation/createActionItem.ts [hunk 8]
```diff
@@ -214,19 +175,45 @@
...
+      },
+      { new: true },
+    ).lean()) as InterfaceEventVolunteerGroup;
+    await EventVolunteer.updateMany(
+      { _id: { $in: newGrp.volunteers } },
+      {
+        $addToSet: { assignments: createActionItem._id },
+      },
+    );
+  }
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/resolvers/Mutation/createActionItem.ts:216]: done

File: src/resolvers/Mutation/createActionItem.ts [hunk 9]
```diff
@@ -60,62 +60,41 @@ export const createActionItem: MutationResolvers["createActionItem"] = async (
...
-    if (currentUserAppProfile !== null) {
-      await cacheAppUserProfile([currentUserAppProfile]);
+  } else if (assigneeType === "EventVolunteerGroup") {
+    assignee = await EventVolunteerGroup.findById(assigneeId).lean();
+    if (!assignee) {
+      throw new errors.NotFoundError(
+        requestContext.translate(EVENT_VOLUNTEER_GROUP_NOT_FOUND_ERROR.MESSAGE),
+        EVENT_VOLUNTEER_GROUP_NOT_FOUND_ERROR.CODE,
+        EVENT_VOLUNTEER_GROUP_NOT_FOUND_ERROR.PARAM,
+      );
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/resolvers/Mutation/createActionItem.ts:94]: not required

File: src/utilities/checks.ts [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+  }
+  if (!currentUserAppProfile) {
+    throw new errors.UnauthorizedError(
+      requestContext.translate(USER_NOT_AUTHORIZED_ERROR.MESSAGE),
+      USER_NOT_AUTHORIZED_ERROR.CODE,
+      USER_NOT_AUTHORIZED_ERROR.PARAM,
+    );
+  }
+  return currentUserAppProfile;
+};
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/utilities/checks.ts:85]: not required

File: src/utilities/checks.ts [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    throw new errors.NotFoundError(
+      requestContext.translate(
+        EVENT_VOLUNTEER_MEMBERSHIP_NOT_FOUND_ERROR.MESSAGE,
+      ),
+      EVENT_VOLUNTEER_MEMBERSHIP_NOT_FOUND_ERROR.CODE,
+      EVENT_VOLUNTEER_MEMBERSHIP_NOT_FOUND_ERROR.PARAM,
+    );
+  }
+  return volunteerMembership;
+};
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/utilities/checks.ts:153]: Building generic entity checker wont help as sometime different population requirements are there

File: src/resolvers/Query/getEventVolunteerGroups.ts [hunk 12]
```diff
@@ -9,14 +15,107 @@
...
+      filteredEventVolunteerGroups = filteredEventVolunteerGroups.filter(
+        (group) => {
+          const tempGroup = group as InterfaceEventVolunteerGroup;
+          const tempLeader = tempGroup.leader as InterfaceUser;
+          const { firstName, lastName } = tempLeader;
+          const name = `${firstName} ${lastName}`.toLowerCase();
+          return name.includes(tempName);
+        },
+      );
+    }
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/resolvers/Query/getEventVolunteerGroups.ts:95]: doesnt change much

File: tests/resolvers/Mutation/updateEventVolunteer.spec.ts [hunk 13]
```diff
@@ -1,200 +1,85 @@
...
+          hasAccepted: false,
+          assignments: [],
+        },
+      },
+      { userId: testUser1?._id.toString() },
+    )) as unknown as InterfaceEventVolunteer;
+    expect(updatedVolunteer.isPublic).toEqual(false);
+    expect(updatedVolunteer.hasAccepted).toEqual(false);
+    expect(updatedVolunteer.assignments).toEqual([]);
   });
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply tests/resolvers/Mutation/updateEventVolunteer.spec.ts:84]: not needed

File: src/resolvers/Mutation/createEventVolunteer.ts [hunk 14]
```diff
@@ -38,77 +34,61 @@
...
+      creator: context.userId,
+      groups: [],
+    });
+
+    // create volunteer membership record
+    await VolunteerMembership.create({
+      volunteer: createdVolunteer._id,
+      event: eventId,
+      status: "invited",
+      createdBy: context.userId,
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/resolvers/Mutation/createEventVolunteer.ts:90]: not needed

File: src/resolvers/Mutation/createEventVolunteer.ts [hunk 15]
```diff
@@ -38,77 +34,61 @@ import type { MutationResolvers } from "../../types/generatedGraphQLTypes";
...
-      eventId: args.data.eventId,
-      groupId: args.data.groupId,
-      isAssigned: false,
-      isInvited: true,
-      creatorId: context.userId,
+      user: userId,
+      event: eventId,
+      creator: context.userId,
+      groups: [],
+    });
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/resolvers/Mutation/createEventVolunteer.ts:83]: not needed

File: tests/resolvers/Query/getVolunteerMembership.spec.ts [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+import type { InterfaceVolunteerMembership } from "../../../src/models";
+import { VolunteerMembership } from "../../../src/models";
+import { getVolunteerMembership } from "../../../src/resolvers/Query/getVolunteerMembership";
+
+export enum MembershipStatus {
+  INVITED = "invited",
+  REQUESTED = "requested",
+  ACCEPTED = "accepted",
+  REJECTED = "rejected",
+}
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply tests/resolvers/Query/getVolunteerMembership.spec.ts:20]: done

File: src/typeDefs/types.ts [hunk 17]
```diff
@@ -252,21 +256,36 @@
...
     creator: User
     event: Event
-    group: EventVolunteerGroup
-    isAssigned: Boolean
-    isInvited: Boolean
-    response: String
-    user: User!
+    groups: [EventVolunteerGroup]
+    hasAccepted: Boolean!
+    isPublic: Boolean!
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/typeDefs/types.ts:270]: name itself is enough to understand the meaning

File: src/models/VolunteerMembership.ts [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+    },
+    updatedBy: {
+      type: Schema.Types.ObjectId,
+      ref: "User",
+    },
+  },
+  {
+    timestamps: true, // Automatically manage `createdAt` and `updatedAt` fields
+  },
+);
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/models/VolunteerMembership.ts:72]: not needed

File: src/models/VolunteerMembership.ts [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+  _id: Types.ObjectId;
+  volunteer: PopulatedDoc<InterfaceEventVolunteer & Document>;
+  group: PopulatedDoc<InterfaceEventVolunteerGroup & Document>;
+  event: PopulatedDoc<InterfaceEvent & Document>;
+  status: "invited" | "requested" | "accepted" | "rejected";
+  createdBy: PopulatedDoc<InterfaceUser & Document>;
+  updatedBy: PopulatedDoc<InterfaceUser & Document>;
+  createdAt: Date;
+  updatedAt: Date;
+}
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/models/VolunteerMembership.ts:23]: createdBy & updatedBy are just used for logging so doesnt affect much

**Review by palisadoes — CHANGES_REQUESTED** [review]
@GlenDsza Try to get coderabbit to approve your changes.

File: schema.graphql [hunk 20]
```diff
@@ -765,20 +776,32 @@
...
+  volunteerUserIds: [ID!]!
   volunteersRequired: Int
 }
 
+enum EventVolunteerGroupOrderByInput {
+  assignments_ASC
+  assignments_DESC
+  members_ASC
+  members_DESC
+}
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply schema.graphql:792]: done

File: schema.graphql [hunk 21]
```diff
@@ -741,21 +748,25 @@ enum EventOrderByInput {
...
   event: Event
-  group: EventVolunteerGroup
-  isAssigned: Boolean
-  isInvited: Boolean
-  response: String
+  groups: [EventVolunteerGroup]
+  hasAccepted: Boolean!
+  hoursHistory: [HoursHistory]
+  hoursVolunteered: Float!
+  isPublic: Boolean!
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply schema.graphql:759]: not needed

File: src/resolvers/Mutation/createActionItem.ts [hunk 22]
```diff
@@ -214,19 +175,41 @@
...
+      assigneeId,
+      { $addToSet: { assignments: createActionItem._id } },
+      { new: true },
+    ).lean()) as InterfaceEventVolunteerGroup;
+
+    await EventVolunteer.updateMany(
+      { _id: { $in: newGrp.volunteers } },
+      { $addToSet: { assignments: createActionItem._id } },
+    );
+  }
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/resolvers/Mutation/createActionItem.ts:212]: EventVolunteer has groups attribute which is an array of object ids. so above workflow might increase time complexity

File: src/typeDefs/mutations.ts [hunk 23]
```diff
@@ -298,9 +302,12 @@ export const mutations = gql`
 
     updateEventVolunteerGroup(
       id: ID!
-      data: UpdateEventVolunteerGroupInput
+      data: UpdateEventVolunteerGroupInput!
     ): EventVolunteerGroup! @auth
 
+    updateVolunteerMembership(id: ID!, status: String!): VolunteerMembership!
+      @auth
+
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/typeDefs/mutations.ts:310]: done

File: src/typeDefs/inputs.ts [hunk 24]
```diff
@@ -637,6 +662,20 @@
     file: String
   }
 
+  input VolunteerMembershipInput {
+    event: ID!
+    group: ID
+    status: String!
+    userId: ID!
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/typeDefs/inputs.ts:669]: done

File: src/typeDefs/inputs.ts [hunk 25]
```diff
@@ -637,6 +662,20 @@
...
+  }
+
+  input VolunteerRankWhereInput {
+    nameContains: String
+    orderBy: String!
+    timeFrame: String!
+    limit: Int
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/typeDefs/inputs.ts:676]: done

File: src/typeDefs/inputs.ts [hunk 26]
```diff
@@ -76,6 +77,7 @@
   }
 
   input ActionItemWhereInput {
+    orgId: ID
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/typeDefs/inputs.ts:80]: it is as intended

File: src/typeDefs/mutations.ts [hunk 27]
```diff
@@ -298,9 +302,12 @@
 
     updateEventVolunteerGroup(
       id: ID!
-      data: UpdateEventVolunteerGroupInput
+      data: UpdateEventVolunteerGroupInput!
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/typeDefs/mutations.ts:305]: it is as intended

File: src/typeDefs/inputs.ts [hunk 28]
```diff
@@ -152,31 +154,53 @@ export const inputs = gql`
...
+
+  input UpdateEventVolunteerInput {
+    assignments: [ID]
+    hasAccepted: Boolean
+    isPublic: Boolean
   }
 
   input UpdateEventVolunteerGroupInput {
-    eventId: ID
+    eventId: ID!
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/typeDefs/inputs.ts:201]: its required

File: src/typeDefs/inputs.ts [hunk 29]
```diff
@@ -152,31 +154,53 @@
...
-    groupId: ID!
+    groupId: ID
+  }
+
+  input EventVolunteerWhereInput {
+    id: ID
+    eventId: ID
+    groupId: ID
+    hasAccepted: Boolean
+    name_contains: String
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/typeDefs/inputs.ts:165]: not required

File: src/resolvers/Query/getVolunteerMembership.ts [hunk 30]
```diff
@@ -0,0 +1,129 @@
...
+    const eventTitleCondition = eventTitle
+      ? membership.event.title.includes(eventTitle)
+      : true;
+
+    const userNameCondition = userName
+      ? (
+          membership.volunteer.user.firstName +
+          membership.volunteer.user.lastName
+        ).includes(userName)
+      : true;
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/resolvers/Query/getVolunteerMembership.ts:91]: every volunteer object has user attribute

File: src/typeDefs/queries.ts [hunk 31]
```diff
@@ -54,17 +60,32 @@
...
+    getVolunteerMembership(
+      where: VolunteerMembershipWhereInput!
+      orderBy: VolunteerMembershipOrderByInput
+    ): [VolunteerMembership]
+
+    getVolunteerRanks(
+      orgId: ID!
+      where: VolunteerRankWhereInput!
+    ): [VolunteerRank]
+
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/typeDefs/queries.ts:88]: it is intended

File: src/typeDefs/queries.ts [hunk 32]
```diff
@@ -54,17 +60,32 @@ export const queries = gql`
...
-      where: EventVolunteerGroupWhereInput
+      where: EventVolunteerGroupWhereInput!
+      orderBy: EventVolunteerGroupOrderByInput
     ): [EventVolunteerGroup]!
 
+    getVolunteerMembership(
+      where: VolunteerMembershipWhereInput!
+      orderBy: VolunteerMembershipOrderByInput
+    ): [VolunteerMembership]
+
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/typeDefs/queries.ts:83]: done

File: src/resolvers/Query/getEventVolunteerGroups.ts [hunk 33]
```diff
@@ -9,14 +15,107 @@
...
+        filteredEventVolunteerGroups = filteredEventVolunteerGroups.sort(
+          (a, b) => a.assignments.length - b.assignments.length,
+        );
+        break;
+      case "assignments_DESC":
+        filteredEventVolunteerGroups = filteredEventVolunteerGroups.sort(
+          (a, b) => b.assignments.length - a.assignments.length,
+        );
+        break;
+    }
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/resolvers/Query/getEventVolunteerGroups.ts:118]: done

File: tests/resolvers/Query/actionItemsByOrganization.spec.ts [hunk 34]
```diff
@@ -1,270 +1,84 @@
...
+      {
+        organizationId: testOrganization?._id,
+        where: {
+          assigneeName: testUser1?.firstName,
+        },
+      },
+      {},
+    )) as unknown as InterfaceActionItem[];
+    expect(actionItems[0].assigneeType).toEqual("User");
+    expect(actionItems[0].assigneeUser.firstName).toEqual(testUser1?.firstName);
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply tests/resolvers/Query/actionItemsByOrganization.spec.ts:82]: Existing testcases are enough for 100% test coverage

File: src/models/ActionItem.ts [hunk 35]
```diff
@@ -5,13 +5,18 @@ import type { InterfaceEvent } from "./Event";
...
 /**
  * Interface representing a database document for ActionItem in MongoDB.
  */
 export interface InterfaceActionItem {
   _id: Types.ObjectId;
-  assignee: PopulatedDoc<InterfaceUser & Document>;
+  assignee: PopulatedDoc<InterfaceEventVolunteer & Document>;
+  assigneeGroup: PopulatedDoc<InterfaceEventVolunteerGroup & Document>;
+  assigneeUser: PopulatedDoc<InterfaceUser & Document>;
+  assigneeType: "EventVolunteer" | "EventVolunteerGroup";
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/models/ActionItem.ts]: Added

**Review by tasneemkoushar — APPROVED** [review]

**Pull request merged**